modules/cli/src/main/scala/scala/cli/commands/scalafix/Scalafix.scala

```diff
@@ -0,0 +1,197 @@
+import scala.cli.commands.{ScalaCommand, SpecificationLevel, compile}
+import scala.cli.config.Keys
+import scala.cli.util.ArgHelpers.*
+import scala.cli.util.ConfigDbUtils
+import scala.collection.mutable
+import scala.collection.mutable.Buffer
+import scala.jdk.CollectionConverters.*
+import scala.jdk.OptionConverters.*
+
+object Scalafix extends ScalaCommand[ScalafixOptions] {
+  override def group: String = HelpCommandGroup.Main.toString
+  override def sharedOptions(options: ScalafixOptions): Option[SharedOptions] = Some(options.shared)
+  override def scalaSpecificationLevel: SpecificationLevel = SpecificationLevel.EXPERIMENTAL
+
+  val hiddenHelpGroups: Seq[HelpGroup] =
+    Seq(
+      HelpGroup.Scala,
+      HelpGroup.Java,
+      HelpGroup.Dependency,
+      HelpGroup.ScalaJs,
+      HelpGroup.ScalaNative,
+      HelpGroup.CompilationServer,
+      HelpGroup.Debug
+    )
+  override def helpFormat: HelpFormat = super.helpFormat
+    .withHiddenGroups(hiddenHelpGroups)
+    .withHiddenGroupsWhenShowHidden(hiddenHelpGroups)
+    .withPrimaryGroup(HelpGroup.Format)
+  override def names: List[List[String]] = List(
+    List("scalafix")
+  )
+
+  override def runCommand(options: ScalafixOptions, args: RemainingArgs, logger: Logger): Unit = {
+    val buildOptions = buildOptionsOrExit(options)
+    val buildOptionsWithSemanticDb = buildOptions.copy(scalaOptions =
+      buildOptions.scalaOptions.copy(semanticDbOptions =
+        buildOptions.scalaOptions.semanticDbOptions.copy(generateSemanticDbs = Some(true))
+      )
+    )
+    val inputs        = options.shared.inputs(args.all).orExit(logger)
+    val threads       = BuildThreads.create()
+    val compilerMaker = options.shared.compilerMaker(threads)
+    val configDb      = ConfigDbUtils.configDb.orExit(logger)
+    val actionableDiagnostics =
+      options.shared.logging.verbosityOptions.actions.orElse(
+        configDb.get(Keys.actions).getOrElse(None)
+      )
+
+    val (sourcePaths, workspace, _) =
+      if (args.all.isEmpty)
+        (Seq(os.pwd), os.pwd, None)
+      else {
+        val s = inputs.sourceFiles().collect {
+          case sc: Script          => sc.path
+          case sc: SourceScalaFile => sc.path
+        }
+        (s, inputs.workspace, Some(inputs))
+      }
+
+    val scalaVersion =
+      options.buildOptions.orExit(logger).scalaParams.orExit(logger).map(_.scalaVersion)
+        .getOrElse(Constants.defaultScalaVersion)
+
+    val scalaBinaryVersion = scalaVersion match
+      case v if v.startsWith("2.12.") => "2.12"
+      case v if v.startsWith("2.13.") => "2.13"
+      case v if v.startsWith("3.")    => "2.13"
+      case _ =>
+        logger.error("Unsupported scala version " + scalaVersion)
+        sys.exit(1)
+    val configFilePathOpt = options.scalafixConf.map(os.Path(_, os.pwd))
+    val relPaths          = sourcePaths.map(_.toNIO.getFileName)
+
+    val scalafix = ScalafixInterface
+      .fetchAndClassloadInstance(scalaBinaryVersion)
```

Comment:
Unsurprisingly, this fails when using the native-image bundle because of the dynamic nature of scalafix-interfaces.

```
➜  /tmp /Users/brice.jaglin/git/scala/scala-cli/out/cli/3.3.3/base-image/nativeImage.dest/scala-cli --power scalafix Test.scala
The `scalafix` sub-command is experimental
Please bear in mind that non-ideal user experience should be expected.
If you encounter any bugs or have feedback to share, make sure to reach out to the maintenance team at https://github.com/VirtusLab/scala-cli
Error: scalafix.interfaces.ScalafixException: Failed to load 'scalafix-interfaces.properties' to lookup versions
For more details, please see '/private/tmp/.scala-build/stacktraces/1723547209-8851074228942378765.log'
```

```
➜  /tmp cat /private/tmp/.scala-build/stacktraces/1723546008-8191378505818715998.log
scalafix.interfaces.ScalafixException: Failed to load 'scalafix-interfaces.properties' to lookup versions
  scalafix.interfaces.Scalafix.fetchAndClassloadInstance(Scalafix.java:126)
  scalafix.interfaces.Scalafix.fetchAndClassloadInstance(Scalafix.java:85)
  scala.cli.commands.scalafix.Scalafix$.runCommand(Scalafix.scala:100)
  scala.cli.commands.scalafix.Scalafix$.runCommand(Scalafix.scala:56)
  scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:388)
  scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:369)
  caseapp.core.app.CaseApp.main(CaseApp.scala:157)
  scala.cli.commands.ScalaCommand.main(ScalaCommand.scala:354)
  caseapp.core.app.CommandsEntryPoint.main(CommandsEntryPoint.scala:169)
  scala.cli.ScalaCliCommands.main(ScalaCliCommands.scala:126)
  scala.cli.ScalaCli$.main0(ScalaCli.scala:295)
  scala.cli.ScalaCli$.main(ScalaCli.scala:119)
  scala.cli.ScalaCli.main(ScalaCli.scala)
java.lang.NullPointerException: inStream parameter is null
  java.base@17.0.6/java.util.Objects.requireNonNull(Objects.java:233)
  java.base@17.0.6/java.util.Properties.load(Properties.java:407)
  scalafix.interfaces.Scalafix.fetchAndClassloadInstance(Scalafix.java:124)
  scalafix.interfaces.Scalafix.fetchAndClassloadInstance(Scalafix.java:85)
  scala.cli.commands.scalafix.Scalafix$.runCommand(Scalafix.scala:100)
  scala.cli.commands.scalafix.Scalafix$.runCommand(Scalafix.scala:56)
  scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:388)
  scala.cli.commands.ScalaCommand.run(ScalaCommand.scala:369)
  caseapp.core.app.CaseApp.main(CaseApp.scala:157)
  scala.cli.commands.ScalaCommand.main(ScalaCommand.scala:354)
  caseapp.core.app.CommandsEntryPoint.main(CommandsEntryPoint.scala:169)
  scala.cli.ScalaCliCommands.main(ScalaCliCommands.scala:126)
  scala.cli.ScalaCli$.main0(ScalaCli.scala:295)
  scala.cli.ScalaCli$.main(ScalaCli.scala:119)
  scala.cli.ScalaCli.main(ScalaCli.scala)
```

We could probably fix these [property lookups](https://github.com/scalacenter/scalafix/blob/c02ef92dcaa839b2e11812ee81f3f7029750b6b9/scalafix-interfaces/src/main/java/scalafix/interfaces/Scalafix.java#L120-L132) by [telling GraalVM about it](https://www.graalvm.org/latest/reference-manual/native-image/dynamic-features/Resources/), but the [classloading just after](https://github.com/scalacenter/scalafix/blob/c02ef92dcaa839b2e11812ee81f3f7029750b6b9/scalafix-interfaces/src/main/java/scalafix/interfaces/Scalafix.java#L135-L136) would fail anyway.

I am afraid we will need to run Scalafix through the scalafix-cli launcher (like scalafmt apparently?) instead of scalafix-interfaces. It would be a bit more cumbersome to pass CLI args, and a couple of advanced features wouldn't be available, but it should work overall. Or maybe there is a way to classload Scalafix within bloop somehow @tgodzik ? It would have the benefit of keeping classloaders warm across invocations.

Anyway, to unlock this PR, I guess we could add an early return if the native-image runtime is detected, and handle native-image compatibility in a later PR? I assume the native-image bundle is the most popular distribution channel though... 😢

Reply:
Ouch! You're right, it seems that it won't make a lot of sence if it  doens't work in native image.
I'll try to replace interfaces by scalafix-cli launcher and see how it is going to turn out.

Reply:
we switched on running scalafix-cli as a process - now native works fine too